### PR TITLE
Add EPID support in PRI.

### PR DESCRIPTION
**Pull request by johneasterday**

### CODE REVIEW

File: storage/storage-samples/storage-di-sample/src/main/java/org/fido/iot/storage/DiDbManager.java [hunk 1]
```diff
@@ -26,14 +26,14 @@ public void createTables(DataSource dataSource) {
 
       String sql = "CREATE TABLE IF NOT EXISTS "
           + "MT_DEVICES("
-          + "GUID CHAR(36) PRIMARY KEY, "
-          + "SERIAL_NO VARCHAR(255), "
+          + "GUID CHAR(36), "
+          + "SERIAL_NO VARCHAR(255) PRIMARY KEY, "
```
Comment (prateekcr) [inline storage/storage-samples/storage-di-sample/src/main/java/org/fido/iot/storage/DiDbManager.java:30]: Component sample Device sends the serial number as '0', everytime, which makes it non-unique for now. Can we please revert this, and do this when device is updated?

File: protocol/src/main/java/org/fido/iot/protocol/Const.java [hunk 2]
```diff
@@ -189,12 +189,10 @@
   //EPID SigInfo Types
   public static final int SG_EPIDv10 = 90; // Intel EPID, version 1.0
   public static final int SG_EPIDv11 = 91; // Intel EPID, version 1.1
-  public static final int SG_EPIDv20 = 92; // Intel EPID, version 2.0
```
Comment (Darshini-Parikh93) [inline protocol/src/main/java/org/fido/iot/protocol/Const.java:192]: Could you please help me understand why EPID20 constants are removed? Will we be supporting only EPID11 in this implementation?

Comment (johneasterday) [reply protocol/src/main/java/org/fido/iot/protocol/Const.java:192]: Yes, that is exactly correct - we no longer support EPID 2.0.

File: storage/storage-samples/storage-di-sample/src/main/java/org/fido/iot/storage/DiDbManager.java (hunk 1 above)
Comment (johneasterday) [reply storage/storage-samples/storage-di-sample/src/main/java/org/fido/iot/storage/DiDbManager.java:30]: If a device reruns DI with the same serial no (as it should) then the entry in this table will be replaced with an updated guid, a new row should not be added if the serial no is the same. Can you explain the use case where you see an issue?

Comment (prateekcr) [reply storage/storage-samples/storage-di-sample/src/main/java/org/fido/iot/storage/DiDbManager.java:30]: This is not really an issue in any real-world use-case, and, agree with you on the functionality aspect there.
Thought that we'd like to keep the old vouchers around for general debug purposes (say to compare vouchers during dev to fix issues, if needed, considering pri device is for learning purposes as well.. but we can do with DI-client-sample as well).
Having said that, serial no as PK is indeed better for real-world use-cases.

File: service/protocol-samples/http-client-to0-sample/src/main/java/org/fido/iot/sample/To0ClientApp.java [hunk 3]
```diff
@@ -71,6 +72,66 @@
       + "1ek15IbeCI5z7BHea2GZGgaK63cyD15gNA==\n"
       + "-----END EC PRIVATE KEY-----";
 
+  private static String sampleOwnerKeyPem = "-----BEGIN CERTIFICATE-----\n"
```
Comment (prateekcr) [inline service/protocol-samples/http-client-to0-sample/src/main/java/org/fido/iot/sample/To0ClientApp.java:75]: Are these the same keys that are used in component sample Owner? If not, we can think of aligning this at a later point of time.

File: service/protocol-samples/http-server-di-sample/src/main/java/org/fido/iot/sample/DiContextListener.java [hunk 4]
```diff
@@ -46,13 +47,85 @@
       + "MHcCAQEEIJTKW2/54N85RLJu0C5fEkAwQiKqxRqHzx5PUfd/M66UoAoGCCqGSM49\n"
       + "AwEHoUQDQgAELAJwkDKz/BaWq1Wx7PjkR5W5LLIbamgSZeVNUlyFM/t0sMAxAWbv\n"
       + "EbDzKu924TX4as3WVjMmfekysx30PlDGJQ==\n"
-      + "-----END EC PRIVATE KEY-----";
+      + "-----END EC PRIVATE KEY-----\n";
+
+  private static final String mfgKeyPemRSA = "-----BEGIN CERTIFICATE-----\n"
```
Comment (prateekcr) [inline service/protocol-samples/http-server-di-sample/src/main/java/org/fido/iot/sample/DiContextListener.java:52]: Same as above.

File: protocol/src/main/java/org/fido/iot/protocol/CryptoService.java [hunk 5]
```diff
@@ -503,18 +513,30 @@ public Composite encode(PublicKey publicKey, int encType) {
...
 
       }
       break;
+      case Const.PK_ENC_CRYPTO: {
+        RSAPublicKey key = (RSAPublicKey) publicKey;
+        final ByteBuffer mod = ByteBuffer.wrap(key.getModulus().toByteArray());
+        final ByteBuffer exp = ByteBuffer.wrap(key.getPublicExponent().toByteArray());
+
+        Composite pkbody = Composite.newArray()
+                .set(0, mod)
```
Comment (prateekcr) [inline protocol/src/main/java/org/fido/iot/protocol/CryptoService.java:526]: Can use Const.FIRST_KEY instead.

**Review by prateekcr — DISMISSED** [review]

File: protocol/src/main/java/org/fido/iot/protocol/Const.java (hunk 2 above)
Comment (Darshini-Parikh93) [reply protocol/src/main/java/org/fido/iot/protocol/Const.java:192]: ok

File: protocol/src/main/java/org/fido/iot/protocol/CryptoService.java [hunk 6]
```diff
@@ -441,8 +450,8 @@ public int getCompatibleHmacType(PublicKey key) {
    * @return signature
    */
   public Composite getSigInfoB(Composite sigInfoA) {
-    if (null != sigInfoA && sigInfoA.size() > 0
-            && Arrays.asList(Const.SG_EPIDv10, Const.SG_EPIDv11, Const.SG_EPIDv20)
+    if (null != sigInfoA
```
Comment (Darshini-Parikh93) [inline protocol/src/main/java/org/fido/iot/protocol/CryptoService.java:453]: Could you please help me understand why `&& sigInfoA.size() > 0` was removed? 
This condition was added for non-EPID device value of eA. Without this Component Sample Device runs into IndexOutOfBoundException.

File: storage/storage-samples/storage-di-sample/src/main/java/org/fido/iot/storage/DiDbStorage.java [hunk 7]
```diff
@@ -77,36 +75,68 @@ public Composite createVoucher(Object createParams) {
...
+
     UUID guid = UUID.randomUUID();
-    Certificate[] issuerChain = resolver.getCertChain(
-        mstr.getAsNumber(Const.FIRST_KEY).intValue());
 
+    int keyType = mstr.getAsNumber(Const.FIRST_KEY).intValue();
     String serialNo = mstr.getAsString(Const.SECOND_KEY);
 
-    if (mstr.get(Const.FOURTH_KEY) != Optional.empty()) {
+    if (keyType == Const.PK_SECP256R1 || keyType == Const.PK_SECP256R1) {
```
Comment (Darshini-Parikh93) [inline storage/storage-samples/storage-di-sample/src/main/java/org/fido/iot/storage/DiDbStorage.java:114]: Shouldn't one of the two constants be `PK_SECP384R1`?

File: protocol/src/main/java/org/fido/iot/protocol/CryptoService.java (hunk 6 above)
Comment (johneasterday) [reply protocol/src/main/java/org/fido/iot/protocol/CryptoService.java:453]: removed by mistake - will add it back.

File: storage/storage-samples/storage-di-sample/src/main/java/org/fido/iot/storage/DiDbStorage.java (hunk 7 above)
Comment (johneasterday) [reply storage/storage-samples/storage-di-sample/src/main/java/org/fido/iot/storage/DiDbStorage.java:114]: Good catch, one should be 384.

File: service/protocol-samples/http-client-to0-sample/src/main/java/org/fido/iot/sample/To0ClientApp.java (hunk 3 above)
Comment (johneasterday) [reply service/protocol-samples/http-client-to0-sample/src/main/java/org/fido/iot/sample/To0ClientApp.java:75]: Yes.

File: protocol/src/main/java/org/fido/iot/protocol/CryptoService.java (hunk 6 above)
Comment (johneasterday) [reply protocol/src/main/java/org/fido/iot/protocol/CryptoService.java:453]: yes

File: protocol/src/main/java/org/fido/iot/protocol/CryptoService.java [hunk 8]
```diff
@@ -609,11 +641,23 @@ public void verifyMaroePrefix(Composite cose) {
    * @return True if the signature matches otherwise false.
    */
   public boolean verify(PublicKey verificationKey, Composite cose, Composite sigInfoA) {
-    if (null != sigInfoA && sigInfoA.size() > 0
-        && Arrays.asList(Const.SG_EPIDv10, Const.SG_EPIDv11, Const.SG_EPIDv20)
+    if (null != sigInfoA && Arrays.asList(Const.SG_EPIDv10, Const.SG_EPIDv11)
```
Comment (Darshini-Parikh93) [inline protocol/src/main/java/org/fido/iot/protocol/CryptoService.java:644]: Please add back the condition `&& sigInfoA.size() > 0`  here as well. As discussed for line 453 above.

Comment (johneasterday) [reply protocol/src/main/java/org/fido/iot/protocol/CryptoService.java]: done

**Review by Darshini-Parikh93 — APPROVED** [review]

**Review by prateekcr — APPROVED** [review]

**Pull request merged**